Slow disk deletion and resurrection of previously deleted disks

On rack2 that currently runs `acdcbd94a4b734f386272e4a57e55348ab429bda`, deleting 90+ disks took more than 15 minutes. This is substantially slower than what it used to be.

There are also a large nunber of previously deleted disks showing up in `faulted` state:
```
root@oxz_switch0:~# omdb db disks list | grep faulted
note: database URL not specified.  Will search DNS.
note: (override with --db-url or OMDB_DB_URL)
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using database URL postgresql://root@[fd00:1122:3344:109::3]:32221,[fd00:1122:3344:105::3]:32221,[fd00:1122:3344:10b::3]:32221,[fd00:1122:3344:107::3]:32221,[fd00:1122:3344:108::3]:32221/omicron?sslmode=disable
note: database schema version matches expected (136.0.0)
WARN: listing disks: found 500 items (the limit).  There may be more items that were ignored.  Consider overriding with --fetch-limit.
03a8d85b-2b24-4cd0-acfe-4665e621239c 10 GiB   faulted                    deleted-03a8d85b-2b24-4cd0-acfe-4665e621239c                    -                                    
053ffff8-37f4-4243-981a-d6509255463f 10 GiB   faulted                    deleted-053ffff8-37f4-4243-981a-d6509255463f                    -                                    
08dbff70-58e8-4066-a269-79c3eb4f5d84 10 GiB   faulted                    deleted-08dbff70-58e8-4066-a269-79c3eb4f5d84                    -                                    
0a3939fa-0e38-4bfd-90af-56a3aee57286 10 GiB   faulted                    deleted-0a3939fa-0e38-4bfd-90af-56a3aee57286                    -                                    
0b4cddc4-cdf9-4645-9521-7b98bd5d60d5 10 GiB   faulted                    deleted-0b4cddc4-cdf9-4645-9521-7b98bd5d60d5                    -                                    
0d5bd532-bfe1-4472-86ff-7247cdfcadea 10 GiB   faulted                    deleted-0d5bd532-bfe1-4472-86ff-7247cdfcadea                    -                                    
1173bf29-c95e-46e6-ada3-b8d426d1ad59 10 GiB   faulted                    deleted-1173bf29-c95e-46e6-ada3-b8d426d1ad59                    -                                    
11b84367-86f3-490e-b650-afeadd4a0d8a 10 GiB   faulted                    deleted-11b84367-86f3-490e-b650-afeadd4a0d8a                    -                                    
125b61bb-6814-4e51-acbb-03fba2922a61 10 GiB   faulted                    deleted-125b61bb-6814-4e51-acbb-03fba2922a61                    -                                    
16f66d78-547e-4a35-a2d2-bdcf105d19ed 10 GiB   faulted                    deleted-16f66d78-547e-4a35-a2d2-bdcf105d19ed                    -                                    
2086c370-c628-45f5-8866-63d2fdb5851c 10 GiB   faulted                    deleted-2086c370-c628-45f5-8866-63d2fdb5851c                    -                                    
20da6097-e2c2-46c8-a268-92d18c61049f 10 GiB   faulted                    deleted-20da6097-e2c2-46c8-a268-92d18c61049f                    -                                    
29801c82-34df-46ae-b19a-faf068afe00f 10 GiB   faulted                    deleted-29801c82-34df-46ae-b19a-faf068afe00f                    -                                    
29f28827-0047-44f3-9489-7e142a15e0a3 10 GiB   faulted                    deleted-29f28827-0047-44f3-9489-7e142a15e0a3                    -                                    
2afe4eb1-c5eb-4001-85b5-44a2ed4bafdd 10 GiB   faulted                    deleted-2afe4eb1-c5eb-4001-85b5-44a2ed4bafdd                    -                                    
34730fa7-397f-41ab-af8c-8b4f1ed35a5c 10 GiB   faulted                    deleted-34730fa7-397f-41ab-af8c-8b4f1ed35a5c                    -                                    
39426cdb-1ffc-4f3c-ad9a-4c4bcec85950 10 GiB   faulted                    deleted-39426cdb-1ffc-4f3c-ad9a-4c4bcec85950                    -                                    
3e1bb4d6-6902-4b8b-ac06-b8025e567b69 10 GiB   faulted                    deleted-3e1bb4d6-6902-4b8b-ac06-b8025e567b69                    -                                    
3f0d03c7-362c-4ec5-a78f-ba9100b22533 10 GiB   faulted                    deleted-3f0d03c7-362c-4ec5-a78f-ba9100b22533                    -                                    
4340a280-2d7f-4553-9514-db31654928c7 10 GiB   faulted                    deleted-4340a280-2d7f-4553-9514-db31654928c7                    -                                    
446fee40-debc-4530-bbd3-c25f7af98ec0 10 GiB   faulted                    deleted-446fee40-debc-4530-bbd3-c25f7af98ec0                    -                                    
5bc81734-99eb-4a6f-9dc4-8c6d7d66dcad 10 GiB   faulted                    deleted-5bc81734-99eb-4a6f-9dc4-8c6d7d66dcad                    -                                    
64b7c17e-823c-4c68-9af9-c0f58ab37b37 10 GiB   faulted                    deleted-64b7c17e-823c-4c68-9af9-c0f58ab37b37                    -                                    
67fb5eaf-d9b7-4329-9a82-22cb4be50266 10 GiB   faulted                    deleted-67fb5eaf-d9b7-4329-9a82-22cb4be50266                    -                                    
69ba2dfc-b424-4933-9339-990f7eefffd9 10 GiB   faulted                    deleted-69ba2dfc-b424-4933-9339-990f7eefffd9                    -                                    
6a378f6b-15ee-4b18-98de-304793d79c4d 10 GiB   faulted                    deleted-6a378f6b-15ee-4b18-98de-304793d79c4d                    -                                    
6a6b68eb-cc59-4bf4-9b29-4ee8f48ed174 10 GiB   faulted                    deleted-6a6b68eb-cc59-4bf4-9b29-4ee8f48ed174                    -                                    
6e6e3453-10c0-4248-ac9b-354ed142849f 10 GiB   faulted                    deleted-6e6e3453-10c0-4248-ac9b-354ed142849f                    -                                    
70c2d84b-7df3-4f05-af25-f8ffc6bcb298 10 GiB   faulted                    deleted-70c2d84b-7df3-4f05-af25-f8ffc6bcb298                    -                                    
7600b290-c704-426b-b552-87d89decdac8 10 GiB   faulted                    deleted-7600b290-c704-426b-b552-87d89decdac8                    -                                    
7689005a-215a-4fd1-94a2-cceabf2c8308 10 GiB   faulted                    deleted-7689005a-215a-4fd1-94a2-cceabf2c8308                    -                                    
78b79d84-a5aa-431a-b981-4949d1e2b457 10 GiB   faulted                    deleted-78b79d84-a5aa-431a-b981-4949d1e2b457                    -                                    
79c6015f-0066-48ba-a973-0af08c9e525c 10 GiB   faulted                    deleted-79c6015f-0066-48ba-a973-0af08c9e525c                    -                                    
7b47ff9b-ed7d-4914-8f65-14101c59da8c 10 GiB   faulted                    deleted-7b47ff9b-ed7d-4914-8f65-14101c59da8c                    -                                    
800fa6eb-3989-4dec-b5aa-fa08adb077e1 10 GiB   faulted                    deleted-800fa6eb-3989-4dec-b5aa-fa08adb077e1                    -                                    
80b5b484-9f21-4c2f-ac76-8368a0471027 10 GiB   faulted                    deleted-80b5b484-9f21-4c2f-ac76-8368a0471027                    -                                    
8808c192-b0e4-4da6-9055-6229e339c6ba 10 GiB   faulted                    deleted-8808c192-b0e4-4da6-9055-6229e339c6ba                    -                                    
88e55704-865f-4dcb-82e1-e2d689e0cbd1 10 GiB   faulted                    deleted-88e55704-865f-4dcb-82e1-e2d689e0cbd1                    -                                    
8de85d24-fa7d-4b29-b5ad-7e571ba70686 10 GiB   faulted                    deleted-8de85d24-fa7d-4b29-b5ad-7e571ba70686                    -                                    
8e469a41-3f79-4d9f-9bd5-bb5da2361042 10 GiB   faulted                    deleted-8e469a41-3f79-4d9f-9bd5-bb5da2361042                    -                                    
94b99f0a-d21e-4b0e-a960-ced303136376 10 GiB   faulted                    deleted-94b99f0a-d21e-4b0e-a960-ced303136376                    -                                    
9e0e9cc6-d199-4166-ab4a-bbe178b2535c 10 GiB   faulted                    deleted-9e0e9cc6-d199-4166-ab4a-bbe178b2535c                    -                                    
9f90bb38-d942-496b-8f46-3475bdbf06ec 10 GiB   faulted                    deleted-9f90bb38-d942-496b-8f46-3475bdbf06ec                    -                                    
a126c530-3be1-4a85-a051-0e530613a01a 10 GiB   faulted                    deleted-a126c530-3be1-4a85-a051-0e530613a01a                    -                                    
a4ea7d2f-3ae4-45d4-962b-ddb4a1347aa9 10 GiB   faulted                    deleted-a4ea7d2f-3ae4-45d4-962b-ddb4a1347aa9                    -                                    
a7b7c817-5286-4740-b55e-220714886102 10 GiB   faulted                    deleted-a7b7c817-5286-4740-b55e-220714886102                    -                                    
abdbbfd3-375f-4367-a7c8-ff7950f4f8a5 10 GiB   faulted                    deleted-abdbbfd3-375f-4367-a7c8-ff7950f4f8a5                    -                                    
ada8529a-2f9e-416d-964d-b6d0b8cf569e 10 GiB   faulted                    deleted-ada8529a-2f9e-416d-964d-b6d0b8cf569e                    -                                    
ae5b6a64-4c1d-4357-a460-0fd4a5be1c47 10 GiB   faulted                    deleted-ae5b6a64-4c1d-4357-a460-0fd4a5be1c47                    -                                    
aef005e9-0b3c-4de6-9b3c-42c13cea3c13 10 GiB   faulted                    deleted-aef005e9-0b3c-4de6-9b3c-42c13cea3c13                    -                                    
af0adefd-277d-42bd-a74c-e4618356aa53 10 GiB   faulted                    deleted-af0adefd-277d-42bd-a74c-e4618356aa53                    -                                    
b31b46b1-3894-4010-823b-f8b2e09e51f0 10 GiB   faulted                    deleted-b31b46b1-3894-4010-823b-f8b2e09e51f0                    -                                    
b8aac489-acdd-46c7-8bdd-5d40a90476eb 10 GiB   faulted                    deleted-b8aac489-acdd-46c7-8bdd-5d40a90476eb                    -                                    
bb4df6b7-33ea-413d-b5b0-94f9e8b32f61 10 GiB   faulted                    deleted-bb4df6b7-33ea-413d-b5b0-94f9e8b32f61                    -                                    
bf6df41b-4d0b-40ae-a1be-62d659d0cbf3 10 GiB   faulted                    deleted-bf6df41b-4d0b-40ae-a1be-62d659d0cbf3                    -                                    
c453bb66-0be9-40dc-9d76-7cc1e711f1fb 10 GiB   faulted                    deleted-c453bb66-0be9-40dc-9d76-7cc1e711f1fb                    -                                    
c6c49604-916b-4321-85fe-57c166c90e74 10 GiB   faulted                    deleted-c6c49604-916b-4321-85fe-57c166c90e74                    -                                    
cd393eb6-589b-4b5d-820e-094ab16812d1 10 GiB   faulted                    deleted-cd393eb6-589b-4b5d-820e-094ab16812d1                    -                                    
d2ab411e-0b32-4a3f-a556-ea82614e3b1f 10 GiB   faulted                    deleted-d2ab411e-0b32-4a3f-a556-ea82614e3b1f                    -                                    
d58b6a40-b5d9-4074-8288-664685f20bdb 10 GiB   faulted                    deleted-d58b6a40-b5d9-4074-8288-664685f20bdb                    -                                    
d6f43faf-9914-4035-80b8-a398e71bf5e3 10 GiB   faulted                    deleted-d6f43faf-9914-4035-80b8-a398e71bf5e3                    -                                    
d7d49285-ef79-4da8-ae22-1a38e9ccb8e4 10 GiB   faulted                    deleted-d7d49285-ef79-4da8-ae22-1a38e9ccb8e4                    -                                    
df8a57a0-2d54-4cf1-bae6-5b8bcce7572d 10 GiB   faulted                    deleted-df8a57a0-2d54-4cf1-bae6-5b8bcce7572d                    -                                    
e1b15808-1bc8-434f-b4a6-7d471aaabc3d 10 GiB   faulted                    deleted-e1b15808-1bc8-434f-b4a6-7d471aaabc3d                    -                                    
e4535581-6cfa-4fc1-9a82-a95cd4de91ae 10 GiB   faulted                    deleted-e4535581-6cfa-4fc1-9a82-a95cd4de91ae                    -                                    
e6f80295-f8e2-4128-a2a9-d54dcfeb73ec 10 GiB   faulted                    deleted-e6f80295-f8e2-4128-a2a9-d54dcfeb73ec                    -                                    
eae964c4-2411-4c5f-a9bb-79e821f1bb9e 10 GiB   faulted                    deleted-eae964c4-2411-4c5f-a9bb-79e821f1bb9e                    -                                    
ef955dc9-a0dd-431b-b952-794dcaa8f516 10 GiB   faulted                    deleted-ef955dc9-a0dd-431b-b952-794dcaa8f516                    -                                    
f1aec42b-00d9-45a2-a26f-1e1dc572732a 10 GiB   faulted                    deleted-f1aec42b-00d9-45a2-a26f-1e1dc572732a                    -                                    
f40dc7d7-7371-4430-a719-44d2c47207d9 10 GiB   faulted                    deleted-f40dc7d7-7371-4430-a719-44d2c47207d9                    -                                    
fd3a1e0d-8fca-40e0-9f5b-ee431b816176 10 GiB   faulted                    deleted-fd3a1e0d-8fca-40e0-9f5b-ee431b816176                    -                                    
ffc00ade-1084-4094-b040-1fd6733ef67b 10 GiB   faulted                    deleted-ffc00ade-1084-4094-b040-1fd6733ef67b                    -                            
```

fwiw, I'm seeing 500's with regularity on dogfood when I'm using terraform to delete disks:

```
│ Error: Unable to delete disk:
│ 
│ API error: DELETE https://oxide.sys.rack2.eng.oxide.computer/v1/disks/0fa8b206-eb65-464a-8160-739108945ee6
│ ----------- RESPONSE -----------
│ Status: 500 Internal
│ Message: Internal Server Error
│ RequestID: c1cd739c-83cd-4526-aaad-c9fd177552c2
│ ------- RESPONSE HEADERS -------
│ Content-Type: [application/json]
│ X-Request-Id: [c1cd739c-83cd-4526-aaad-c9fd177552c2]
│ Date: [Thu, 10 Apr 2025 14:37:11 GMT]
│ Content-Length: [124]
│ 
╵
make: *** [destroy] Error 1
```

They delete successfully on a re-run.

--- a/nexus/db-queries/src/db/datastore/region.rs
+++ b/nexus/db-queries/src/db/datastore/region.rs
@@ -20,11 +20,11 @@ use crate::db::model::SqlU16;
 use crate::db::model::to_db_typed_uuid;
 use crate::db::pagination::Paginator;
 use crate::db::pagination::paginated;
-use crate::db::queries::region_allocation::RegionParameters;
+use crate::db::queries::region_allocation;
+use crate::db::queries::regions_hard_delete;
 use crate::db::update_and_check::UpdateAndCheck;
 use crate::db::update_and_check::UpdateStatus;
 use async_bb8_diesel::AsyncRunQueryDsl;
-use diesel::dsl::sql_query;
 use diesel::prelude::*;
 use nexus_config::RegionAllocationStrategy;
 use nexus_types::external_api::params;
@@ -258,10 +258,10 @@ impl DataStore {
                 } => (block_size, blocks_per_extent, extent_count),
             };
 
-        let query = crate::db::queries::region_allocation::allocation_query(
+        let query = region_allocation::allocation_query(
             volume_id,
             maybe_snapshot_id,
-            RegionParameters {
+            region_allocation::RegionParameters {
                 block_size,
                 blocks_per_extent,
                 extent_count,
@@ -273,10 +273,10 @@ impl DataStore {
 
         let conn = self.pool_connection_authorized(&opctx).await?;
 
-        let dataset_and_regions: Vec<(CrucibleDataset, Region)> =
-            query.get_results_async(&*conn).await.map_err(|e| {
-                crate::db::queries::region_allocation::from_diesel(e)
-            })?;
+        let dataset_and_regions: Vec<(CrucibleDataset, Region)> = query
+            .get_results_async(&*conn)
+            .await
+            .map_err(|e| region_allocation::from_diesel(e))?;
 
         info!(
             self.log,
@@ -302,47 +302,23 @@ impl DataStore {
         }
 
         let conn = self.pool_connection_unauthorized().await?;
+
         self.transaction_retry_wrapper("regions_hard_delete")
             .transaction(&conn, |conn| {
                 let region_ids = region_ids.clone();
+
                 async move {
                     use nexus_db_schema::schema::region::dsl;
 
-                    // Remove the regions
-                    diesel::delete(dsl::region)
+                    let dataset_ids: Vec<Uuid> = diesel::delete(dsl::region)
                         .filter(dsl::id.eq_any(region_ids))
-                        .execute_async(&conn)
+                        .returning(dsl::dataset_id)
+                        .get_results_async(&conn)
                         .await?;
 
-                    // Update datasets to which the regions belonged.
-                    sql_query(
-                        r#"
-WITH size_used_with_reservation AS (
-  SELECT
-    crucible_dataset.id AS crucible_dataset_id,
-    SUM(
-      CASE
-        WHEN block_size IS NULL THEN 0
-        ELSE
-          CASE
-            WHEN reservation_percent = '25' THEN
-              (block_size * blocks_per_extent * extent_count) / 4 +
-              (block_size * blocks_per_extent * extent_count)
-          END
-      END
-    ) AS reserved_size
-  FROM crucible_dataset
-  LEFT JOIN region ON crucible_dataset.id = region.dataset_id
-  WHERE crucible_dataset.time_deleted IS NULL
-  GROUP BY crucible_dataset.id
-)
-UPDATE crucible_dataset
-SET size_used = size_used_with_reservation.reserved_size
-FROM size_used_with_reservation
-WHERE crucible_dataset.id = size_used_with_reservation.crucible_dataset_id"#,
-                    )
-                    .execute_async(&conn)
-                    .await?;
+                    let query =
+                        regions_hard_delete::dataset_update_query(dataset_ids);
+                    query.execute_async(&conn).await?;
 
                     // Whenever a region is hard-deleted, validate invariants
                     // for all volumes

--- a/nexus/db-queries/src/db/queries/mod.rs
+++ b/nexus/db-queries/src/db/queries/mod.rs
@@ -13,6 +13,7 @@ mod next_item;
 pub mod network_interface;
 pub mod oximeter;
 pub mod region_allocation;
+pub mod regions_hard_delete;
 pub mod sled_reservation;
 pub mod virtual_provisioning_collection_update;
 pub mod vpc;

--- a/nexus/db-queries/src/db/queries/regions_hard_delete.rs
+++ b/nexus/db-queries/src/db/queries/regions_hard_delete.rs
@@ -1,0 +1,60 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Implementation of query to update crucible_dataset size_used after
+//! hard-deleting regions
+
+use crate::db::datastore::RunnableQueryNoReturn;
+use crate::db::raw_query_builder::QueryBuilder;
+use diesel::sql_types;
+use uuid::Uuid;
+
+/// Update the affected Crucible dataset rows after hard-deleting regions
+pub fn dataset_update_query(
+    dataset_ids: Vec<Uuid>,
+) -> impl RunnableQueryNoReturn {
+    let mut builder = QueryBuilder::new();
+
+    builder.sql(
+        "WITH
+  size_used_with_reservation AS (
+    SELECT
+      crucible_dataset.id AS crucible_dataset_id,
+      SUM(
+        CASE
+          WHEN block_size IS NULL THEN 0
+          ELSE
+            CASE
+              WHEN reservation_percent = '25' THEN
+                (block_size * blocks_per_extent * extent_count) / 4 +
+                (block_size * blocks_per_extent * extent_count)
+            END
+        END
+      ) AS reserved_size
+    FROM crucible_dataset
+    LEFT JOIN region ON crucible_dataset.id = region.dataset_id
+    WHERE
+      crucible_dataset.time_deleted IS NULL AND
+      crucible_dataset.id IN (",
+    );
+
+    for (idx, dataset_id) in dataset_ids.into_iter().enumerate() {
+        if idx != 0 {
+            builder.sql(",");
+        }
+        builder.param().bind::<sql_types::Uuid, _>(dataset_id);
+    }
+
+    builder.sql(
+        ")
+    GROUP BY crucible_dataset.id
+  )
+  UPDATE crucible_dataset
+  SET size_used = size_used_with_reservation.reserved_size
+  FROM size_used_with_reservation
+  WHERE crucible_dataset.id = size_used_with_reservation.crucible_dataset_id",
+    );
+
+    builder.query::<()>()
+}

--- a/nexus/db-queries/src/db/queries/regions_hard_delete.rs
+++ b/nexus/db-queries/src/db/queries/regions_hard_delete.rs
@@ -36,15 +36,10 @@ pub fn dataset_update_query(
     LEFT JOIN region ON crucible_dataset.id = region.dataset_id
     WHERE
       crucible_dataset.time_deleted IS NULL AND
-      crucible_dataset.id IN (",
+      crucible_dataset.id = ANY (",
     );
 
-    for (idx, dataset_id) in dataset_ids.into_iter().enumerate() {
-        if idx != 0 {
-            builder.sql(",");
-        }
-        builder.param().bind::<sql_types::Uuid, _>(dataset_id);
-    }
+    builder.param().bind::<sql_types::Array<sql_types::Uuid>, _>(dataset_ids);
 
     builder.sql(
         ")

--- a/nexus/db-queries/src/db/queries/regions_hard_delete.rs
+++ b/nexus/db-queries/src/db/queries/regions_hard_delete.rs
@@ -58,3 +58,49 @@ pub fn dataset_update_query(
 
     builder.query::<()>()
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::db::explain::ExplainableAsync;
+    use crate::db::pub_test_utils::TestDatabase;
+    use crate::db::raw_query_builder::expectorate_query_contents;
+    use omicron_test_utils::dev;
+    use uuid::Uuid;
+
+    // This test is a bit of a "change detector", but it's here to help with
+    // debugging too. If you change this query, it can be useful to see exactly
+    // how the output SQL has been altered.
+    #[tokio::test]
+    async fn expectorate_query() {
+        let query =
+            dataset_update_query(vec![Uuid::nil(), Uuid::nil(), Uuid::nil()]);
+
+        expectorate_query_contents(
+            &query,
+            "tests/output/dataset_update_query.sql",
+        )
+        .await;
+    }
+
+    // Explain the possible forms of the SQL query to ensure that it
+    // creates a valid SQL string.
+    #[tokio::test]
+    async fn explainable() {
+        let logctx = dev::test_setup_log("explainable");
+        let db = TestDatabase::new_with_pool(&logctx.log).await;
+        let pool = db.pool();
+        let conn = pool.claim().await.unwrap();
+
+        let query =
+            dataset_update_query(vec![Uuid::nil(), Uuid::nil(), Uuid::nil()]);
+
+        let _ = query
+            .explain_async(&conn)
+            .await
+            .expect("Failed to explain query - is it valid SQL?");
+
+        db.terminate().await;
+        logctx.cleanup_successful();
+    }
+}

--- a/nexus/db-queries/tests/output/dataset_update_query.sql
+++ b/nexus/db-queries/tests/output/dataset_update_query.sql
@@ -17,7 +17,7 @@ WITH
       FROM
         crucible_dataset LEFT JOIN region ON crucible_dataset.id = region.dataset_id
       WHERE
-        crucible_dataset.time_deleted IS NULL AND crucible_dataset.id IN ($1, $2, $3)
+        crucible_dataset.time_deleted IS NULL AND crucible_dataset.id = ANY ($1)
       GROUP BY
         crucible_dataset.id
     )

--- a/nexus/db-queries/tests/output/dataset_update_query.sql
+++ b/nexus/db-queries/tests/output/dataset_update_query.sql
@@ -1,0 +1,31 @@
+WITH
+  size_used_with_reservation
+    AS (
+      SELECT
+        crucible_dataset.id AS crucible_dataset_id,
+        sum(
+          CASE
+          WHEN block_size IS NULL THEN 0
+          ELSE CASE
+          WHEN reservation_percent = '25'
+          THEN block_size * blocks_per_extent * extent_count / 4
+          + block_size * blocks_per_extent * extent_count
+          END
+          END
+        )
+          AS reserved_size
+      FROM
+        crucible_dataset LEFT JOIN region ON crucible_dataset.id = region.dataset_id
+      WHERE
+        crucible_dataset.time_deleted IS NULL AND crucible_dataset.id IN ($1, $2, $3)
+      GROUP BY
+        crucible_dataset.id
+    )
+UPDATE
+  crucible_dataset
+SET
+  size_used = size_used_with_reservation.reserved_size
+FROM
+  size_used_with_reservation
+WHERE
+  crucible_dataset.id = size_used_with_reservation.crucible_dataset_id

